File: pkg/toulmin/detect_run_cycle.go
//ff:func feature=engine type=validator control=sequence
//ff:what detectRunCycle — DFS over RunGraph edges to reject cyclic execution composition
package toulmin

import "fmt"

// detectRunCycle walks the graph-of-graphs reachable from root via each node's
// RunGraph edge, keyed by *Graph pointer identity, using a 3-color DFS
// (0=unvisited, 1=visiting, 2=done). Re-entering a visiting graph is an execution
// cycle and returns an error; a shared sub-graph reached by two paths (diamond DAG)
// is legal thanks to the done color. Execution composition must be a DAG.
func detectRunCycle(root *Graph) error {
	state := make(map[*Graph]int)
	var visit func(g *Graph) error
	visit = func(g *Graph) error {
		if state[g] == 1 {
			return fmt.Errorf("toulmin: run cycle detected at graph %q", g.name)
		}
		if state[g] == 2 {
			return nil
		}
		state[g] = 1
		for i := range g.rules {
			sub := g.rules[i].RunGraph
			if sub == nil {
				continue
			}
			if err := visit(sub); err != nil {
				return err
			}
		}
		state[g] = 2
		return nil
	}
	return visit(root)
}

File: ts/src/detect-run-cycle.ts
import type { Graph } from "./graph.js";   // 타입 전용(순환 안전)

// detectRunCycle walks the graph-of-graphs reachable from root via each node's
// runGraph edge, keyed by Graph object identity, using a 3-color DFS
// (WHITE=unvisited, GRAY=visiting, BLACK=done). Re-entering a GRAY graph is an
// execution cycle; a shared sub-graph reached by two paths (diamond DAG) is legal
// thanks to BLACK. Execution composition must be a DAG.
export function detectRunCycle(root: Graph): Error | null {
  const WHITE = 0, GRAY = 1, BLACK = 2;
  const color = new Map<Graph, number>();   // ★ Graph 객체 정체성으로 키잉
  function dfs(g: Graph): Error | null {
    color.set(g, GRAY);
    for (const r of g.rules) {
      const sub = r.runGraph;
      if (!sub) continue;
      const c = color.get(sub) ?? WHITE;
      if (c === GRAY) return new Error(`toulmin: run cycle detected at graph "${sub.name}"`);
      if (c === WHITE) { const err = dfs(sub); if (err) return err; }
    }
    color.set(g, BLACK);
    return null;
  }
  return dfs(root);
}
